raft/log_test.go

```diff
@@ -546,36 +499,33 @@ func TestCompaction(t *testing.T) {
 	}
 
 	for i, tt := range tests {
-		func() {
-			defer func() {
-				if r := recover(); r != nil {
-					if tt.wallow {
-						t.Errorf("%d: allow = %v, want %v: %v", i, false, true, r)
-					}
-				}
-			}()
-
+		t.Run(fmt.Sprint(i), func(t *testing.T) {
 			storage := NewMemoryStorage()
-			for i := uint64(1); i <= tt.lastIndex; i++ {
-				storage.Append([]pb.Entry{{Index: i}})
-			}
-			raftLog := newLog(storage, raftLogger)
-			raftLog.maybeCommit(tt.lastIndex, 0)
-			raftLog.appliedTo(raftLog.committed)
-
-			for j := 0; j < len(tt.compact); j++ {
-				err := storage.Compact(tt.compact[j])
-				if err != nil {
-					if tt.wallow {
-						t.Errorf("#%d.%d allow = %t, want %t", i, j, false, tt.wallow)
+			if !tt.wallow {
+				require.Panics(t, func() {
+					for j := 0; j < len(tt.compact); j++ {
+						_ = storage.Compact(tt.compact[j])
```

Comment:
I think we should not change the test case when we refactor.
We should prepare the log before `Compact`.

It should be like this, if we have to use `require.Panics`.

```go
storage := NewMemoryStorage()
for i := uint64(1); i <= tt.lastIndex; i++ {
    storage.Append([]pb.Entry{{Index: i}})
}
raftLog := newLog(storage, raftLogger)
raftLog.maybeCommit(tt.lastIndex, 0)
raftLog.appliedTo(raftLog.committed)

verifyIn := require.NoPanics
if !tt.wallow {
   verifyIn = require.Panics
}

verifyIn(t, func() {
   for j := 0; j < len(tt.compact); j++ {
	err := storage.Compact(tt.compact[j])
	if err != nil {
		require.False(t, tt.wallow)
		continue
	}

	require.Equal(t, tt.wleft[j], len(raftLog.allEntries()))
    }
})
```

Reply:
Just my two cents, use `defer` looks more clear than using `Panics` 😂

Reply:
good ieda, let me resolve that